### ISSUE TITLE
Stackoverflowexception when tracing Finalizer

### ISSUE THREAD

**Issue by GetCodeTrack** [dotnet/coreclr#14763]
When tracing the sample below the profiler gets in an infinite loop (keeps entering in the EnterMethod callback). Eventually resulting in a stackoverflow exception.

If you remove the finalizer, everything works fine. Tried it on my own profiler (www.getcodetrack.com), but also on Jetbrains Dottrace. Same result. 
In .NET 4.0 and 2.0 it works fine.

```
static void Main(string[] args)
        {
            for (int i = 0; i < 999999; i += 1)
            {
                var thing = new testClass();
                Console.WriteLine("Created thing {0}", i);
            }
 
            Console.Write("completed");
 
            Console.ReadKey();
        }
 
        public class testClass
        {
            private int[] ints;
 
            public testClass()
            {
                this.ints = Enumerable.Range(0, 99).ToArray();
            }
           
            ~testClass()
            {
 
            }
        }
```

**Comment by jkotas** [comment]
cc @sywhang

**Comment by sywhang** [comment]
Hi @GetCodeTrack, thanks for reporting this issue. While I'm investigating this issue, I had some trouble repro the issue with the sample you provided. It would be great if you could tell us a little bit more about your current setup to help us investigate this issue!

- Are you using the latest master branch of coreclr to implement your profiler? 
- If not, which version of .NET Core are you using?
- Which profiler flags have you set while implementing your profiler?
- Are there any environment variables you are setting?

**Comment by GetCodeTrack** [comment]
Hi @sywhang ,

I'm really sorry but I don't seem to be able to reproduce it again either. I might have been doing something terribly wrong before.

Nevertheless, I do get an exception., and I think I know the reason:
Whenever ThreadCreated goes off I create a thread static context (using TLS). 
If I get in the EnterMethod I try to use this context, which normally is ok since a thread created has always been triggered on that thread.

In .NET Core (2.0) I see that the ThreadCreated for the finalizer thread does not get called, which results in the context not being initialized. The consequence is a crash because the EnterMethod of the finalizer finds an illegal (uninitialized) context.

If you change the for loop to 999 instead of 999999 all is fine, because the finalizer doesn't get called before tracing is over...

I don't know if its still relevant but my eventmask was 0x00011240.
I tried with the .NET Core 2.0 SDK.

Gr,
Nico

**Comment by sywhang** [comment]
Hi @GetCodeTrack,

That's completely fine. Thanks for verifying that for me. 

Would there be any chance that your issue is related to #13499 ? If so, this issue has been fixed with this: https://github.com/dotnet/coreclr/pull/13663, and this fix should've been included in the 2.0.2 servicing release. 

Could you please try downloading the latest .NET Core SDK (https://www.microsoft.com/net/download/windows), and see if the issue still persists? 

Cheers,
Sung

**Comment by GetCodeTrack** [comment]
Hi @sywhang ,

That ( #13499 ) sounds exactly like my case. I installed 2.0.2 with the link you gave but the issue is still there.
Dotnet -version returned 2.0.2 so it is installed correctly I guess...

Gr,
Nico

**Comment by sywhang** [comment]
Hi @GetCodeTrack, 

I quickly tried re-running the tests I made while verifying the fix for #13499, and it seems like the profiler is still getting ThreadCreated callbacks on finalizer threads. I'll try to take a deeper look later tonight and get back to you.

In the meantime can you please make sure you're using the correct version of the SDK?

**Comment by sywhang** [comment]
Hi @GetCodeTrack,

Sorry, I directed you to a wrong path. I meant you should download the 2.0.2 servicing version of the **runtime**, not the SDK. Just installing the 2.0.2 version of the SDK does not mean that you have 2.0.2 version of the runtime. You can download the official daily build of servicing here: https://github.com/dotnet/core-setup/blob/release/2.0.0/README.md#officially-released-builds

Once you install the latest runtime, please make sure it creates a directory that looks like ```C:\Program Files\dotnet\shared\Microsoft.NETCore.App\2.0.4-servicing-25831-01``` (the actual number may look different, but anything >2.0.2 should be fine). To see if you are actually using the latest runtime, check that you have the correct coreclr.dll module being loaded when you run your managed code by looking at the path of the coreclr.dll and seeing if it's from the directory made from installing the latest runtime.

Please let me know if that works!

Thanks,
Sung

**Comment by fbuyvolkot** [comment]
Hi @sywhang 

Can you please tell how to specify a specific runtime to be used by a process?
After installing 2.0.4 runtime, coreclr.dll still loads from C:\Program Files\dotnet\shared\Microsoft.NETCore.App\2.0.0

Thanks

**Comment by sywhang** [comment]
Hi @fbuyvolkot ,

You could try adding the ```<RuntimeFrameworkVersion></RuntimeFrameworkVersion>``` tag to your .csproj file, and specify the runtime version to ```2.0.4-servicing-24831-01``` or whichever runtime version you've installed. 

An alternative hack is to just replace the coreclr.dll in the 2.0.0 directory although I wouldn't recommend doing that.

**Comment by chrisbrook83** [comment]
Hi @sywhang 

I'm trying to test out the `servicing` release too, from here:
https://github.com/dotnet/core-setup/blob/release/2.0.0/README.md#officially-released-builds

When I set up the project with the appropriate RuntimeFrameworkVersion, it no longer builds with the familiar error:
```The type 'Object' is defined in an assembly that is not referenced. You must add a reference to assembly 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'```

I tried adding the netstandard nuget package though, and it still didn't work. Any suggestions?

**Comment by GetCodeTrack** [comment]
hi @sywhang ,

I finally found the time to test CodeTrack again with 2.0.4-servicing-25831-01.
Just running dotnet.exe is loading 2.0.0 as before and setting<RuntimeFrameworkVersion></RuntimeFrameworkVersion> in the csproj file gives me build errors.

The good news though: if I replace the coreclr.dll in the 2.0.0 (like you told us not to :-)), the profiler works fine.

So I guess the fix is good, we just need to find a way to use the correct version...

Gr,
Nico

**Comment by sywhang** [comment]
@GetCodeTrack @chrisbrook83 

Sorry about that -- I repro'd your issues and found a quick workaround -- you can rename the directory ```C:\Program Files\dotnet\shared\Microsoft.NETCore.App\2.0.4-servicing-24831-01``` to simply ```C:\Program Files\dotnet\shared\Microsoft.NETCore.App\2.0.4``` (and please also remove the ```<RuntimeFrameworkVersion>``` tag in your .csproj file too!) 
Then it should be loading the coreclr.dll from 2.0.4 directory. I tried doing this myself just now and seems like it works, but if it doesn't please feel free to comment again.

Cheers,
Sung

**Comment by cbranch-bhg** [comment]
I tried copying 2.0.4 servicing to 2.0.4 directory and can confirm this appears to fix the issue. It seems like a temporary workaround however a more permanent one would be nice.

**Comment by sywhang** [comment]
Hey @cbranch-bhg ,

I'm not an expert on how the dotnet CLI interacts with the runtime or how it loads the proper runtime. You might want to file an issue / question in the dotnet cli repo here: https://github.com/dotnet/cli to ask for more permanent fix or getarounds for that.

Since the original issue seems unreproduceable, and the issue regarding coreclr has already been fixed in a previous issue (#13499), I will close this issue here.
